crates/editor/src/signature_help_popover.rs

```diff
@@ -0,0 +1,553 @@
+) -> Option<(String, Vec<(Range<usize>, MarkdownHighlight)>)> {
+    let (signature_information, maybe_active_signature, maybe_active_parameter) = (
+        signature_help.signatures,
+        signature_help.active_signature,
+        signature_help.active_parameter,
+    );
+
+    let function_options_count = signature_information.len();
+
+    let maybe_signature_information = maybe_active_signature
+        .and_then(|active_signature| signature_information.get(active_signature as usize));
+    let signature_information = if let Some(signature_information) = maybe_signature_information {
+        Some(signature_information)
+    } else {
+        signature_information.first()
+    }?;
+
+    let str_for_join = ", ";
+    let parameter_length = signature_information
+        .parameters
+        .as_ref()
+        .map(|parameters| parameters.len())
+        .unwrap_or(0);
+    let mut highlight_start = 0;
+    let (markdown, mut highlights): (Vec<_>, Vec<_>) = signature_information
+        .parameters
+        .as_ref()?
+        .iter()
+        .enumerate()
+        .filter_map(|(i, parameter_information)| {
+            let string = match parameter_information.label.clone() {
+                lsp::ParameterLabel::Simple(string) => string,
+                lsp::ParameterLabel::LabelOffsets(offset) => signature_information
+                    .label
+                    .chars()
+                    .skip(offset[0] as usize)
+                    .take((offset[1] - offset[0]) as usize)
+                    .collect::<String>(),
+            };
+            let string_length = string.len();
+
+            let result = if let Some(active_parameter) = maybe_active_parameter {
+                if i == active_parameter as usize {
+                    Some((
+                        string,
+                        Some((
+                            highlight_start..(highlight_start + string_length),
+                            SIGNATURE_HELP_HIGHLIGHT,
+                        )),
+                    ))
+                } else {
+                    Some((string, None))
+                }
+            } else {
+                Some((string, None))
+            };
+
+            if i != parameter_length {
+                highlight_start += string_length + str_for_join.len();
+            }
+
+            result
+        })
+        .unzip();
+    let markdown = markdown.join(str_for_join);
+    let markdown = if function_options_count >= 2 {
+        let suffix = format!("(+{} overload)", function_options_count - 1);
+        let highlight_start = markdown.len() + 1;
+        highlights.push(Some((
+            highlight_start..(highlight_start + suffix.len()),
+            SIGNATURE_HELP_OVERLOAD_HIGHLIGHT,
+        )));
+        let markdown = format!("{} {}", markdown, suffix);
+        markdown
+    } else {
+        markdown
+    };
+
+    let highlights = highlights.into_iter().flatten().collect::<Vec<_>>();
```

Comment:
NITs: 

*
```
.skip(offset[0] as usize)
                    .take((offset[1] - offset[0]) as usize)
```
seems to work but overall would be safer to avoid potential panics from `[]` (use `.get`) and `-` (use `.saturating_sub` or comparisons).
This is an external data coming from "untrusted" LSP servers, so we could get in some nonsense in theory and panicking on that seems too excessive.

Also, that `.clone()` in the related `match` is only needed for one of the match arms, so we can stop extra clones.

* `let highlights = highlights.into_iter().flatten().collect::<Vec<_>>();`
we can allocate that only when needed, in `else`.

*
```
let markdown = format!("{} {}", markdown, suffix);
        markdown
```
could be just a `format!("{markdown} {suffix}")` one-liner

Reply:
@SomeoneToIgnore 
> seems to work but overall would be safer to avoid potential panics from [] (use .get) and - (use .saturating_sub or comparisons).
This is an external data coming from "untrusted" LSP servers, so we could get in some nonsense in theory and panicking on that seems too excessive.

Thank you for your reviewing.
It might be unclear from the file changes tab on GitHub, but offset is of type `[u32; 2]`, so the safety of accessing the 0th and 1st elements should be guaranteed at compile time. Therefore, I believe it is fine. However, if it still concerns you, I am willing to respect your opinion.

> Also, that .clone() in the related match is only needed for one of the match arms, so we can stop extra clones

Since the label’s lifetime will end at the end of the function, we need to create an owned String instead of using &String or &str

> let highlights = highlights.into_iter().flatten().collect::<Vec<_>>();
we can allocate that only when needed, in else.

Thank you so match. This is my fault. I fixed it on 543821ff4 and a7a85158e